### ISSUE TITLE
Mocking Virtual Variadic Template Functions

### ISSUE THREAD

**Issue by szegel** [google/googletest#3166]
Any Ideas how to mock something like this?

```c++
class Test {
public:
    virtual template <typename A, typename ...B, typename ...Args>
    OtherTestClass<A(B...)> testFunction(A& obj, Args&&... args) = 0;
};
```

**Comment by sbenzaquen** [comment]
That is not legal C++. A virtual function can't be a template also.
Sharing real working code would help us help you.

Note that `MOCK_METHOD` does not support templates in general yet.
If you want to mock templates you can do so via indirection.

Eg

```
struct Foo {
  MOCK_METHOD(void, BarImpl, (int));
  MOCK_METHOD(void, BarImpl, (double));

  // Supported only for int and double.
  template <typename T>
  void Bar(T t) {
    BarImpl(t);
  }
};
```

**Comment by szegel** [comment]
My bad, and thank you for clarifying this.  I have an example at work that has a similar interface to this (can't share or I get in big trouble).  I spent some time trying to mock it before I realized it wasn't virtual.  The real working code is a variadic template member function of a abstract base class.  This got me curious if it would be mock-able if it was virtual.  As you pointed out, the answer is no b/c the language doesn't allow this behavior.

Thanks!
